doc/vlan_stacking/vlan_stacking_HLD.md

```diff
@@ -0,0 +1,862 @@
+  - [Configuration](#configuration)
+- [High-Level Design](#high-level-design)
+  - [VlanMgr](#vlanmgr)
+  - [DB](#db)
+    - [CONFIG_DB](#config_db)
+    - [APPL_DB](#appl_db)
+  - [SAI API](#sai-api)
+- [Configuration and management](#configuration-and-management)
+  - [CLI/YANG model Enhancements](#cliyang-model-enhancements)
+    - [CLI](#cli)
+    - [Yang model](#yang-model)
+  - [Examples](#examples)
+    - [Examples of QinQ Configuration](#examples-of-qinq-configuration)
+    - [Examples of VLAN Translation Configuration](#examples-of-vlan-translation-configuration)
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Restrictions/Limitations](#restrictionslimitations)
+- [Testing Requirements/Design](#testing-requirementsdesign)
+  - [System Test Cases](#system-test-cases)
+- [Future Work](#future-work)
+
+# List of Tables
+
+* [Table 1: Revision](#revision)
+* [Table 2: Definitions/Abbreviations](#definitionsabbreviations)
+* [Table 3: SAI attributes related to VLAN stacking](#sai-api)
+
+# Revision
+
+###### Table 1: Revision
+
+| Rev   |   Date     | Author        | Change Description                 |
+|:-----:|:----------:|:-------------:|------------------------------------|
+| 0.1   |20-Dec-2021 | Tommy Tseng   | Initial version                    |
+| 0.2   |26-Jan-2022 | Tommy Tseng   | Update CLI command                 |
+| 0.2   |23-Feb-2022 | Tommy Tseng   | Modify DB schema                   |
+| 0.3   |25-Feb-2022 | Yeh Jun-ying  | Update example, Yang model and CLI |
+| 0.4   |23-Feb-2022 | Tommy Tseng   | Update Yang model and spec         |
+# Scope
+
+This document describes the high level design of VLAN stacking feature.
+
+# Definitions/Abbreviations
+
+###### Table 2: Definitions/Abbreviations
+
+| Abbreviation | Full form              |
+|--------------|------------------------|
+| VID          | Virtual LAN Identifier |
+| S-VLAN       | Service Provider VLAN  |
+| C-VLAN       | Customer VLAN          |
+| S-TAG        | S-VLAN tag             |
+| C-TAG        | C-VLAN tag             |
+
+# Overview
+
+This document describes the design details of the VLAN stacking feature. VLAN stacking is a feature designed for service providers who carry traffic of multiple customers across  their networks and are required to maintain the VLAN and Layer 2 protocol configurations of each customer without impacting the traffic of other customers. VLAN stacking include QinQ feature and VLAN Translation feature.
+
+IEEE 802.1Q tunneling (QinQ tunneling) uses a single Service Provider VLAN (SPVLAN) for customers who have multiple VLANs. Customer VLAN IDs are preserved and traffic from different customers is segregated within the service provider’s network even when they use the same customer-specific VLAN IDs. QinQ tunneling expands VLAN space by using a VLAN-in-VLAN hierarchy, preserving the customer’s original tagged packets, and adding SPVLAN tags to each frame (also called double tagging).
```

Comment:
Are we planning to make any changes in the kernel for VLAN stacking/translation?

Reply:
Currently, no, is there any concern for change the kernel code ?

Reply:
Generally, we create an entry in the kernel for things we program in the HW e.g VLAN, in this case, we assume that we don't need VLAN stacking support for CPU originated traffic, correct?

Reply:
@venkatmahalingam Yes, VLAN stacking do not need for CPU traffic.